Sources/NIO/ByteBuffer-extract.swift

```diff
@@ -0,0 +1,191 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the SwiftNIO open source project
+//
+// Copyright (c) 2017-2018 Apple Inc. and the SwiftNIO project authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See CONTRIBUTORS.txt for the list of SwiftNIO project authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+/// A base ByteBuffer "Slicer", this base one just splits by a separator.
+public struct ByteBufferSliceSplitIterator: IteratorProtocol {
+    
+    var byteBuffer: ByteBuffer
+    var separator: UInt8
+    let start: Int
+    let length: Int
+    
+    private var currentIndex: Int
+    
+    public typealias Element = ByteBuffer
+    mutating public func next() -> ByteBuffer? {
+        
+        return byteBuffer.withVeryUnsafeBytes { pointer -> ByteBuffer? in
+            
+            let startPoint   = pointer.baseAddress!.assumingMemoryBound(to: UInt8.self)
+            
+            var address      = startPoint.advanced(by: currentIndex)
+            let finalAddress =    address.advanced(by: length)
+            
+            let initialPointer = address
+
+            while address < finalAddress && address.pointee != 0 {
+                
+                if address.pointee == separator {
+                    break
+                }
+                address = address.advanced(by: 1)
+            }
```

Comment:
This entire loop can be better handled by using an `UnsafeBufferPointer` and looping over that. It'll avoid some of the manual checking of indices.

Reply:
How? and why? I didn't use any check inside, the only check is in the `init`

Reply:
You're checking pointer addresses against lengths. You can just do a `for` loop over the contents of a `Collection` of UTF-8.

Reply:
You can still use an `UnsafeBufferPointer` here for this.